experimental/parser/parse_type.go

```diff
@@ -0,0 +1,196 @@
+	// repeated.foo needs to be interpreted as repeated .foo.
+	var (
+		mods   []token.Token
+		tyPath ast.Path
+	)
+	for !c.Done() && tyPath.Nil() {
+		next := c.Peek()
+		if !canStartPath(next) {
+			break
+		}
+
+		tyPath = parsePath(p, c)
+		if tyPath.Absolute() {
+			break // Absolute paths cannot start with a modifier, so we are done.
+		}
+
+		first, _ := iters.First(tyPath.Components)
+		ident := first.AsIdent()
+		if ident.Nil() {
+			break // If this starts with an extension, we're done.
+		}
+
+		// Here is a nasty case. Suppose the user has written within message
+		// scope something like
+		//
+		//  package .foo.bar = 5;
+		//
+		// This is a syntax error but only because the name of the field is a
+		// non-trivial path. However, we would like for this to be diagnosed as
+		// a package declaration. Thus, if this looks like a bad package, we
+		// break it up so that the type is "package" and the path is ".foo.bar",
+		// so that the DeclPackage code path can diagnose it.
+		//
+		// We require that this have no modifiers and that it not be followed by
+		// a path, so that the following productions are *not* treated as weird
+		// packages:
+		//
+		//  optional package.foo.bar;
+		//  package.foo.bar baz;
+		//
+		// Note that this does not apply inside of type lists. This is because
+		// type lists *only* contain types, and not productions started by
+		// keywords.
+		//
+		// This case applies to many other keywords. See cannotStartType.
+		if !isList && len(mods) == 0 &&
+			cannotStartType(ident.Text()) &&
+			!canStartPath(c.Peek()) {
+			kw, path := tyPath.Split(1)
+			if !path.Nil() {
+				return ast.TypePath{Path: kw}.AsAny(), path
+			}
+		}
+
+		// Check if ident is a modifier, and if so, peel it off.
+		//
+		// We need to be careful to only peel off `stream` inside of a method
+		// type. If the entire path is a single identifier, we always peel it
+		// off, since code that follows handles turning it back into a path
+		// based on what comes after it.
+		var isMod bool
+		_, rest := tyPath.Split(1)
+		switch ast.TypePrefixByName(ident.Name()) {
+		case ast.TypePrefixOptional, ast.TypePrefixRepeated, ast.TypePrefixRequired:
+			isMod = !isList || rest.Nil()
+		case ast.TypePrefixStream:
+			isMod = isInMethod || rest.Nil()
+		}
+
+		if isMod {
+			mods = append(mods, ident)
+			tyPath = rest
+		}
+	}
+
+	if tyPath.Nil() {
+		if len(mods) == 0 {
+			return ast.TypeAny{}, ast.Path{}
+		}
+
+		// Pop the last mod and make that into the type path. This makes
+		// `optional optional` work as a type.
+		last := mods[len(mods)-1]
+		tyPath = astx.NewPath(p.Context, last, last)
+		mods = mods[:len(mods)-1]
+	}
+
+	ty := ast.TypePath{Path: tyPath}.AsAny()
+
+	// Next, look for some angle brackets. We need to do this before draining
+	// mods, because angle brackets bind more tightly than modifiers.
+	if angles := c.Peek(); angles.Text() == "<" && !angles.IsLeaf() {
```

Comment:
If angle brackets generate a sub-tree of tokens, how do we anticipate handling comparative operators (`<`, `>`, `<=`, `>=`) in a future where we directly support CEL expressions (or any kind of expressions really)? Seems like the lexer would have to analyze them as singular tokens and then this part of the parser would need to fuse a sequence of tokens into a sub-tree for cases where it expects `<` to be the start of a group like this (and in message literals).

Reply:
That is a very "we'll burn that bridge when we get there" scenario. It is extremely obnoxious when languages use `<` as both a bracket and an infix operator. Every language deals with this in some way and all of them are kind of bad.

I think the result will be that `<` will not be a token tree in the future. But again, that is very "we'll burn that bridge when we get there".